Assigning masked view to a masked view fails

for example, I want to get {{0,11},{0,13},{0,15}} from

```
xt::xarray<size_t> data = {{0,1},
        {2,3},
        {4,5}};
```

I'm doing assigning like that:
```
xt::xarray<size_t> new_data = xt::zeros<size_t>(data.shape());
xt::xarray<bool> col_mask = {false, true};

for(size_t i=0; i<data.shape()[0]; i++){ //iterate over rows
	auto row_masked = xt::masked_view(xt::view(data, i, xt::all()), col_mask);
	auto new_row_masked = xt::masked_view(xt::view(data_new, i, xt::all()), col_mask);
	row_masked += 10;
	new_row_masked = row_masked;
}
```

but compiler says:

> /usr/include/xtensor/xbroadcast.hpp:168:73: error: no type named ‘bool_load_type’ in ‘using xexpression_type = std::decay_t<const xt::xmasked_view<xt::xview<xt::xarray_container<xt::uvector<int>, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, long unsigned int, xt::xall<long unsigned int> >, xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&>&> {aka class xt::xmasked_view<xt::xview<xt::xarray_container<xt::uvector<int>, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, long unsigned int, xt::xall<long unsigned int> >, xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&>}’
[build]          using bool_load_type = typename xexpression_type::bool_load_type;

how can I achieve desired result?

also I tried to assign one view to another element by element and got this:

> xample.cpp: In function ‘int main()’:
example.cpp:21:36: error: ambiguous overload for ‘operator=’ (operand types are ‘xt::xaccessible<xt::xmasked_view<xt::xview<xt::xarray_container<xt::uvector<long unsigned int, std::allocator<long unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, long unsigned int, xt::xall<long unsigned int> >, xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&> >::reference {aka xtl::xmasked_value<long unsigned int&, bool&>}’ and ‘xt::xaccessible<xt::xmasked_view<xt::xview<xt::xarray_container<xt::uvector<long unsigned int, std::allocator<long unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, long unsigned int, xt::xall<long unsigned int> >, xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&> >::reference {aka xtl::xmasked_value<long unsigned int&, bool&>}’)
    new_row_masked[i] = row_masked[i];
                                    ^
In file included from /usr/include/xtensor/xmasked_view.hpp:13:0,
                 from example.cpp:2:
/usr/include/xtl/xmasked_value.hpp:92:9: note: candidate: xtl::xmasked_value<T, B>& xtl::xmasked_value<T, B>::operator=(const T1&) [with T1 = xtl::xmasked_value<long unsigned int&, bool&>; T = long unsigned int&; B = bool&]
         DEFINE_ASSIGN_OPERATOR(=);
         ^
/usr/include/xtl/xmasked_value.hpp:92:9: note: candidate: xtl::xmasked_value<T, B>& xtl::xmasked_value<T, B>::operator=(const xtl::xmasked_value<T1, B1>&) [with T1 = long unsigned int&; B1 = bool&; T = long unsigned int&; B = bool&]
         DEFINE_ASSIGN_OPERATOR(=);
         ^
In file included from /usr/include/xtensor/xmasked_view.hpp:13:0,
                 from example.cpp:2:
/usr/include/xtl/xmasked_value.hpp:30:11: note: candidate: xtl::xmasked_value<long unsigned int&, bool&>& xtl::xmasked_value<long unsigned int&, bool&>::operator=(const xtl::xmasked_value<long unsigned int&, bool&>&) <deleted>
     class xmasked_value
           ^~~~~~~~~~~~~
/usr/include/xtl/xmasked_value.hpp:30:11: note: candidate: xtl::xmasked_value<long unsigned int&, bool&>& xtl::xmasked_value<long unsigned int&, bool&>::operator=(xtl::xmasked_value<long unsigned int&, bool&>&&) <deleted>

wow, it's not even possible to assign "row_masked" masked_view from my pastebin example to an xarray. I hoped that I can do so and copy elements by hand, but no:

> /usr/include/xtensor/xassign.hpp:247:63: error: ‘const class xt::xmasked_view<xt::xview<xt::xarray_container<xt::uvector<long unsigned int, std::allocator<long unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, long unsigned int, xt::xall<long unsigned int> >, xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&>’ has no member named ‘has_linear_assign’; did you mean ‘minus_assign’?
                    (e1.layout() != layout_type::dynamic && e2.has_linear_assign(e1.strides()));

--- a/include/xtensor/xmasked_view.hpp
+++ b/include/xtensor/xmasked_view.hpp
@@ -116,6 +116,8 @@ namespace xt
         using size_type = typename inner_types::size_type;
         using difference_type = typename data_type::difference_type;
 
+        using bool_load_type = xtl::xmasked_value<typename data_type::bool_load_type, mask_type>;
+
         using shape_type = typename data_type::shape_type;
         using strides_type = typename data_type::strides_type;
 
@@ -194,6 +196,9 @@ namespace xt
         template <class It>
         const_reference element(It first, It last) const;
 
+        template <class S>
+        bool has_linear_assign(const S& strides) const noexcept;
+
         data_type& value() noexcept;
         const data_type& value() const noexcept;
 
@@ -218,6 +223,8 @@ namespace xt
         const_stepper stepper_begin(const S& shape) const noexcept;
         template <class S>
         const_stepper stepper_end(const S& shape, layout_type l) const noexcept;
+
+        self_type& operator=(const self_type& rhs);
 
         template <class E>
         self_type& operator=(const xexpression<E>& e);
@@ -487,6 +494,13 @@ namespace xt
     }
     //@}
 
+    template <class CTD, class CTM>
+    template <class S>
+    inline bool xmasked_view<CTD, CTM>::has_linear_assign(const S& strides) const noexcept
+    {
+        return m_data.has_linear_assign(strides) && m_mask.has_linear_assign(strides);
+    }
+
     /**
      * Return an expression for the values of the xmasked_view.
      */
@@ -549,6 +563,13 @@ namespace xt
     inline auto xmasked_view<CTD, CTM>::stepper_end(const S& shape, layout_type l) const noexcept -> const_stepper
     {
         return const_stepper(value().stepper_end(shape, l), visible().stepper_end(shape, l));
+    }
+
+    template <class CTD, class CTM>
+    inline auto xmasked_view<CTD, CTM>::operator=(const self_type& rhs) -> self_type&
+    {
+        temporary_type tmp(rhs);
+        return this->assign_temporary(std::move(tmp));
     }
 
     template <class CTD, class CTM>

--- a/test/test_xmasked_view.cpp
+++ b/test/test_xmasked_view.cpp
@@ -242,4 +242,20 @@ namespace xt
                                     { 0.7, 8. , 0.9}};
         EXPECT_EQ(data, expected1);
     }
+
+    TEST(xmasked_view, view)
+    {
+	xt::xarray<size_t> data = {{0,1}, {2,3}, {4,5}};
+	xt::xarray<size_t> data_new = xt::zeros<size_t>(data.shape());
+	xt::xarray<bool> col_mask = {false, true};
+
+	auto row_masked = xt::masked_view(xt::view(data, 0, xt::all()), col_mask);
+        auto new_row_masked = xt::masked_view(xt::view(data_new, 0, xt::all()), col_mask);
+
+        row_masked += 10;
+        new_row_masked = row_masked;
+
+        EXPECT_EQ(data_new(0, 0), size_t(0));
+        EXPECT_EQ(data_new(0, 1), size_t(11));
+    }
 }

